Redis-wire protocol

m4ttheux opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add an opt-in Redis wire-protocol server backed by `Join` tables: a configured `redis.port` serves `GET`/`MGET` and `HGET`/`HMGET` point lookups (plus `AUTH`, `SELECT`, `PING`, `ECHO`) against ClickHouse tables mapped to Redis database numbers.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

m4ttheux committed:
REDIS PROTOCOL initial implementation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80353) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80353) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: alexey-milovidov<br/>:x: m4ttheux<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=80353) it.</sub>

alexey-milovidov commented:
Needs a test and a demo.

m4ttheux committed:
extended config, added a clickhouse-redis mapping layer, expanded supported commands, changed file structure

m4ttheux committed:
fix lint, add tests

alexey-milovidov commented on tests/integration/test_redis_protocol/general_commands.py:
```diff
@@ -0,0 +1,42 @@
+    value = redis.ping()
+    assert value == "PONG"
+
+    value = redis.echo("Hello")
```
Would expect a little bit more.

alexey-milovidov replied:
The test was reworked and extended: it now actually installs the `redis` server config (it previously passed `main_configs=[]`, so the section was never loaded) and covers `PING`/`ECHO`/`QUIT`, `SELECT` error cases, `GET`/`MGET` on a string database, `HGET`/`HMGET` on a hash database, missing keys returning Nil, and type/column error cases. See the summary comment on the PR for a demo transcript.

alexey-milovidov replied:
🕵 The test has been rewritten since then (it is `tests/integration/test_redis_protocol/test.py` now — the old one never actually started a server, because it passed `main_configs=[]`). It now covers `PING`, `ECHO` (including an empty bulk string), `QUIT`, `SELECT` of an unconfigured database, `AUTH` (with and without a user name, wrong password, and a user without the `SELECT` grant), `GET`/`MGET`/`HGET`/`HMGET` including missing keys and stored empty strings, non-string value types, unsupported `Join` variants, value types that cannot represent a missing key, `LowCardinality` values, row policies, re-resolution of the table and of the mapping after DDL and `SYSTEM RELOAD CONFIG`, oversized RESP length prefixes, and that an error keeps the connection in sync — 14 tests in total. Happy to add more if something is still missing.

m4ttheux committed:
impelemnt get, mget, hget, hmget; add tests

m4ttheux committed:
fix lint

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into REDIS-PROTOCOL

# Conflicts:
#	programs/server/Server.cpp
#	src/Interpreters/ClientInfo.cpp
#	src/Interpreters/ClientInfo.h
#	src/Storages/IStorage.h

alexey-milovidov committed:
Adapt Redis wire protocol support to current master APIs

- \`TCPServerConnectionFactory\`: override \`createConnectionImpl\` instead of the removed virtual \`createConnection\`.
- \`Server.cpp\`: use \`socketBindListen(server_settings, ...)\`, \`makeServerParams\` and \`connection_filter\` for the Redis \`TCPServer\`, matching other protocol servers.
- \`setThreadName\` now takes a \`ThreadName\` enum: add \`REDIS_HANDLER\` entry.
- \`IColumn::getDataAt\` now returns \`std::string_view\`: construct \`String\` explicitly.
- \`getInMemoryMetadataPtr\` now takes \`(context, bypass_metadata_cache)\` and returns a \`StorageMetadataHandle\` that must be bound to an lvalue.
- \`SessionLog\`: bump the \`Interface\` enum count static_assert for the new \`REDIS\` value (placed after \`ARROW_FLIGHT\` which claimed value 10 on master).
- \`ServerType::shouldStop\`: fix the port name to \`redis.port\` to match the config key used in \`Server.cpp\`.

Also fix bugs found while reworking:

- \`RedisHandler::run\` busy-looped forever on \`while (in->eof());\` after a client disconnected; now the connection loop exits on EOF.
- HMGET returned only the last requested field, because \`joinGet\` returns a single column; \`StorageJoin::getChunkByKeys\` now performs one \`joinGet\` per requested column and returns a multi-column chunk.
- A missing column in \`getChunkByKeys\` was silently skipped, leading to a logical error on an empty chunk; now it throws \`NO_SUCH_COLUMN_IN_TABLE\`.
- The RESP reader truncated bulk strings at the first \`\r\`; now it reads exactly the declared length, so keys and values are binary-safe. This made the \`readStringUntilCR\`/\`skipToCRLFOrEOF\` additions to \`ReadHelpers\` unnecessary — reverted.
- \`supportsGetRequests\` is checked once when the database mapping is initialized instead of on every request.
- Remove the unused \`enable_ssl\` config setting (TLS is not implemented) and the dead \`getValueByKey\` method.
- Comment out the \`redis\` section in the default \`config.xml\`/\`config.yaml.example\` instead of listening on port 9006 with a placeholder table mapping by default.
- Style: Allman braces, angle-bracket includes, \`LoggerPtr\`/\`getLogger\`, camelCase method names.

Related: https://github.com/ClickHouse/ClickHouse/pull/80353

alexey-milovidov committed:
Rework test_redis_protocol integration test

The test never installed its config: it passed main_configs=[] so the
\`redis\` section was not loaded, and the string mapping lacked the
mandatory \`value_column\`. Install a minimal \`redis.xml\` via main_configs,
and extend the coverage: PING/ECHO/QUIT, SELECT error cases, GET/MGET on
a string database, HGET/HMGET on a hash database, missing keys returning
Nil, type mismatches and unknown columns returning errors.

Related: https://github.com/ClickHouse/ClickHouse/pull/80353

alexey-milovidov commented:
I brought this branch up to date with the current `master` (it was ~13 months stale) and reworked it to build against the current APIs: `TCPServerConnectionFactory::createConnectionImpl`, `setThreadName(ThreadName)`, `IColumn::getDataAt` returning `std::string_view`, `getInMemoryMetadataPtr(context, bypass_metadata_cache)`, and the new `socketBindListen`/`makeServerParams`/`connection_filter` server registration. `Interface::REDIS` moved to value 11 because `ARROW_FLIGHT` claimed 10 on `master` in the meantime.

While reworking, I fixed several bugs:

- `RedisHandler::run` busy-looped on `while (in->eof());` after a client disconnect, burning a full CPU core per closed connection.
- `HMGET` returned only the last requested field: `joinGet` returns a single column, so `StorageJoin::getChunkByKeys` now performs one `joinGet` per requested column.
- A missing column was silently skipped in `getChunkByKeys`, causing an exception on the empty result; now it throws `NO_SUCH_COLUMN_IN_TABLE` cleanly.
- The RESP reader truncated bulk strings at the first `\r`; it now reads exactly the declared length (binary-safe keys/values).
- `ServerType::shouldStop` used the port name `redis_port` while the server registers `redis.port`, so `SYSTEM STOP LISTEN` could not match it.
- The integration test never installed its config (`main_configs=[]`), and the sample `string` mapping lacked the mandatory `value_column` — the test could not have passed as written. It now installs a minimal `redis.xml` and covers `PING`/`ECHO`/`QUIT`, `SELECT` error cases, `GET`/`MGET`, `HGET`/`HMGET`, missing keys returning Nil, and type/column error cases.
- The default `config.xml` opened port 9006 with a placeholder table mapping on every server; the section is now a commented example, like other optional protocol ports.

Demo against a locally built server with a `Join` table mapped to Redis databases 0 (hash) and 1 (string):

```
OK  PING: 'PONG'
OK  ECHO: b'Hello world'
OK  ECHO binary-safe: b'a\rb\nc'
OK  GET before SELECT: Exception('ERR Redis db not set')
OK  SELECT out of range: Exception('ERR DB index is out of range')
OK  SELECT 0 (hash): 'OK'
OK  HGET city: b'New York'
OK  HGET surname: b'Smith'
OK  HMGET: [b'Johnson', b'Los Angeles']
OK  HGET missing key -> nil: None
OK  GET on hash db: Exception('ERR GET command can only be applied to a database of type string')
OK  HGET unknown column: Exception('ERR There is no column no_such_column in table default.name_surname_map')
OK  SELECT 1 (string): 'OK'
OK  GET: b'Smith'
OK  MGET: [b'Smith', b'Johnson', b'Brown']
OK  GET missing key -> nil: None
OK  HGET on string db: Exception('ERR HGET command can only be applied to a database of type hash')
OK  COMMAND: 'OK'
OK  CLIENT SETINFO: 'OK'
OK  AUTH: 'OK'
OK  QUIT: 'OK'

21/21 checks passed

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into REDIS-PROTOCOL

alexey-milovidov committed:
Fix clang-tidy: initialize `MapDescription::db_type`

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80353&sha=2307909f5372d6c99f39e7edc2027e7ef1d971e2&name_0=PR&name_1=Build%20%28arm_tidy%29
https://github.com/ClickHouse/ClickHouse/pull/80353

alexey-milovidov committed:
Address review: authentication, Nil vs empty string, value serialization, mapping validation

- `AUTH` is now real: `AUTH password` authenticates the `default` user and
  `AUTH username password` authenticates the given user via `Session`.
  Data commands (`SELECT`/`GET`/`MGET`/`HGET`/`HMGET`) require an authenticated
  session; without an explicit `AUTH`, the first data command implicitly tries
  the `default` user with an empty password. All catalog lookups and reads run
  through the authenticated session context, and every lookup checks the
  `SELECT` privilege on the involved columns.
- A missing key is now reported as RESP Nil while a stored empty string is
  returned as an empty bulk string: `getChunkByKeys` was replaced with
  `getBlockByKeys`, which requests Nullable result columns (the same mechanism
  as `joinGetOrNull`), so "not found" is an explicit NULL instead of being
  conflated with the default value. `ECHO ''` also returns an empty bulk string.
- Values are serialized with the column type's text serialization instead of
  `IColumn::getDataAt`, so non-string columns (e.g. `UInt8`) are returned in
  their text representation instead of raw memory bytes.
- The database mapping is validated on first use: the table must have exactly
  one key column, it must match the configured `key_column`, the key must be
  `String`/`FixedString`, and `value_column` must exist for string databases.

The integration test covers the new behavior: explicit and implicit
authentication, wrong password, a user without the `SELECT` grant,
empty-vs-Nil distinction, and non-string value serialization.

alexey-milovidov committed:
Add documentation for the Redis wire protocol interface

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,300 @@
+void RedisHandler::run()
+{
+    setThreadName(ThreadName::REDIS_HANDLER);
+    session = std::make_unique<Session>(server.context(), ClientInfo::Interface::REDIS);
```
Fixed in 452e8fb01d4: `AUTH` now authenticates a ClickHouse user via `Session` (`AUTH password` for the `default` user, `AUTH username password` for any user). Data commands require an authenticated session; without an explicit `AUTH`, the first data command implicitly tries the `default` user with an empty password (matching an unprotected Redis server), and fails if that user has a password. All catalog lookups and reads go through the authenticated session context, and every lookup checks the `SELECT` privilege on the involved columns. Covered by `test_auth` in the integration test (wrong password, RBAC denial for a user without the grant).

alexey-milovidov commented on src/Server/RedisProtocolResponse.h:
```diff
@@ -0,0 +1,100 @@
+    void serialize(WriteBuffer & out) final
+    {
+        Writer writer(out);
+        if (str.empty())
```
Fixed in 452e8fb01d4: the lookup layer now returns Nullable columns (the same mechanism as `joinGetOrNull`), so a missing key is an explicit NULL serialized as Nil, while a stored empty string (and `ECHO ""`) is serialized as an empty bulk string. Covered by the integration test (a row with empty values vs a missing key).

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,300 @@
+            auto table = redis_db->getTable();
+            auto value_column = std::static_pointer_cast<RedisProtocol::RedisStringMapping>(redis_db)->getValueColumnName();
+            auto result_chunk = table->getChunkByKeys({get_request.getKey()}, {value_column}, server.context());
+            auto result = String(result_chunk.getColumns()[0]->getDataAt(0));
```
Fixed in 452e8fb01d4: values are now rendered with the column type's text serialization (`ISerialization::serializeText`) instead of `IColumn::getDataAt`, so non-string columns are returned in their text representation (e.g. `UInt8` `30` as `"30"`), and a missing value of any type is Nil via the Nullable lookup result. Covered by the integration test (`age UInt8` field).

alexey-milovidov commented on src/Server/RedisHandlerFactory.cpp:
```diff
@@ -0,0 +1,83 @@
+        if (description.clickhouse_table.empty())
+            throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER, "clickhouse_table is not set for Redis database {}", key);
+
+        description.key_column = server.config().getString(fmt::format("redis.db.{}.key_column", key));
```
Fixed in 452e8fb01d4: the mapping is validated when the database is first selected — the table must have exactly one key column, it must match the configured `key_column`, the key column must be `String`/`FixedString` (Redis keys are strings), and `value_column` must exist for string databases. A composite-key `Join` table or a mistyped `key_column` is now rejected with a clear error at `SELECT` time instead of a confusing `Mismatched number of keys` error at the first lookup.

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into REDIS-PROTOCOL

# Conflicts:
#	src/Interpreters/SessionLog.cpp

alexey-milovidov committed:
Fix ppc64le build: `ECHO` collides with the `<termios.h>` macro

On ppc64le glibc, `<termios.h>` defines `ECHO` as a macro, so the
`Command::ECHO` constant and the `CommandType::ECHO` enumerator failed to
compile with `expected unqualified-id`. Rename both to `ECHO_COMMAND`.

CI report: https://s3.amazonaws.com/clickhouse-test-reports/PRs/80353/39cc1900694651010e1cf70bddccadc6578479b5/build_ppc64le/build_clickhouse/build_clickhouse.log
Related: https://github.com/ClickHouse/ClickHouse/pull/80353

alexey-milovidov committed:
Address review: resolve the mapped table per request, bound RESP lengths, reject unsupported `Join` variants

- The Redis database mapping cached a `StoragePtr`, so a connected session kept
  reading from an old `Join` object after `DROP TABLE`, `RENAME TABLE`, or
  drop-and-recreate of the configured table. The mapping now stores only the
  `StorageID`; the table is resolved and validated for every request.
- The RESP array length and bulk-string length are fully client-controlled and
  were used for `reserve` and `String::resize` before anything was read. Both are
  now bounded (by the Redis protocol limits `PROTO_MAX_MULTIBULK_LEN` and
  `proto-max-bulk-len`), so an oversized length header is rejected up front.
- `StorageJoin::supportsGetRequests` returned `true` for every `Join` table, while
  `joinGet` only supports `ANY LEFT`. A mapping to `Join(ALL, LEFT, ...)` passed
  `SELECT` and failed only on the first lookup; it is now rejected by `SELECT`.

The integration test covers all three, and the documentation states the
`ANY LEFT` requirement and the per-request table resolution.

Related: https://github.com/ClickHouse/ClickHouse/pull/80353

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,405 @@
+                    "There is no value_column {} configured for Redis database {} in table {}",
+                    mapping.value_column, db_, mapping.clickhouse_table);
+
+            redis_clickhouse_mapping[db_]
```
Fixed in `fc63c43fbd7`. The Redis database mapping now stores only the `StorageID` (`RedisProtocolMapping.h`), and `RedisHandler::resolveTable` resolves the table through `DatabaseCatalog` and re-validates it on every request, so `DROP TABLE`, `RENAME TABLE`, and drop-and-recreate are visible to already connected clients. Covered by `test_table_is_resolved_for_every_request`.

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,405 @@
+            ErrorCodes::INVALID_CONFIG_PARAMETER,
+            "Table {} does not exist in database {}", mapping.clickhouse_table, mapping.clickhouse_db);
+
+    if (!table_ptr->supportsGetRequests())
```
Fixed in `fc63c43fbd7`. `StorageJoin::supportsGetRequests` now returns `true` only for `kind == Left` with `strictness == Any`/`RightAny`, which is exactly what `joinGet` implements, so a mapping to `Join(ALL, LEFT, ...)` or `Join(ANY, INNER, ...)` is rejected by `SELECT` instead of failing on the first lookup. Covered by `test_unsupported_join_variant`, and the documentation now states the `ANY LEFT` requirement.

alexey-milovidov commented on src/Server/RedisProtocolReaderWriter.h:
```diff
@@ -0,0 +1,142 @@
+                ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT,
+                "Tried to read RESP type {} as a bulk string", static_cast<char>(type));
+
+        auto size = readInteger();
```
Fixed in `fc63c43fbd7`. The bulk-string length is now checked against `MAX_BULK_STRING_SIZE` (64 MiB, the Redis `proto-max-bulk-len` default order of magnitude) before `String::resize`, so an oversized length header is rejected without allocating anything. Covered by `test_oversized_length_prefixes`.

alexey-milovidov commented on src/Server/RedisProtocolRequest.h:
```diff
@@ -0,0 +1,354 @@
+                ErrorCodes::UNKNOWN_PACKET_FROM_CLIENT,
+                "Cannot parse incoming request. Unexpected RESP type: {}", static_cast<char>(type));
+
+        command_len = reader.readInteger();
```
Fixed in `fc63c43fbd7`. The RESP array length is now checked against `MAX_ARRAY_SIZE` (1048576, the Redis `PROTO_MAX_MULTIBULK_LEN`) in `RedisRequest::deserialize`, before any per-command parser can use it for `reserve` or loops. Covered by `test_oversized_length_prefixes`.

alexey-milovidov committed:
Merge branch 'master' into REDIS-PROTOCOL

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into REDIS-PROTOCOL

alexey-milovidov committed:
Address review: reread the mapping on `SYSTEM RELOAD CONFIG`, one table snapshot per Redis command

Two correctness gaps reported by the AI review on the previous head:

1. The Redis database mapping was parsed once, when the listener was created, and every
   connection reused that snapshot. `SYSTEM RELOAD CONFIG` therefore had no effect while the
   port stayed the same, so a removed or retargeted mapping kept exposing the old table.
   `RedisHandlerFactory` now only validates the configuration at startup, and `RedisHandler`
   reads the mapping of the selected database from the live server configuration for every
   request (`RedisProtocol::parseDBDescription`). The `RedisClickHouseMapping` class hierarchy
   is gone: the freshly read `MapDescription` carries everything the handler needs.

2. `getBlockByKeys` requested every column with a separate `joinGet` call, and `MGET` looped
   over the keys, taking and releasing the `StorageJoin` read lock each time - so a single
   `MGET` or `HMGET` could mix values from different states of the table. `getBlockByKeys` now
   takes a list of keys, builds a multi-row key block, and looks up all the keys and all the
   columns under a single read lock (in batches of `DEFAULT_BLOCK_SIZE`, because `joinGet`
   returns at most `max_joined_block_rows` rows at once). `MGET` issues one request for all
   its keys.

Tests: `test_mapping_is_reread_after_reload_config` retargets the mapping of Redis database 4
to another table, reloads the configuration, and checks that both an already connected client
and a new one see the new table, and that removing the mapping stops exposing it;
`test_mget_returns_values_of_all_keys` checks the order and the Nil/empty distinction of a
multi-key `MGET`. All 10 tests of `test_redis_protocol` pass locally.

alexey-milovidov commented on src/Server/RedisHandlerFactory.cpp:
```diff
@@ -0,0 +1,83 @@
+RedisHandlerFactory::RedisHandlerFactory(IServer & server_)
+    : server(server_), log(getLogger("RedisHandlerFactory"))
+{
+    config = std::make_shared<RedisProtocol::Config>();
```
🕵 Fixed in `7a56eace474`. The listener no longer caches the mapping: `RedisHandlerFactory` only validates the `redis.db` section at startup (so a mistake is still reported immediately), and `RedisHandler` reads the mapping of the selected database out of the live `server.config()` for every request via the new `RedisProtocol::parseDBDescription`. `SYSTEM RELOAD CONFIG` is therefore visible to already connected clients: a retargeted mapping starts serving the new table, and a removed one fails with `Redis database N is not configured`. Covered by `test_mapping_is_reread_after_reload_config`.

alexey-milovidov commented on src/Storages/StorageJoin.cpp:
```diff
@@ -929,6 +931,60 @@ class JoinSource final : public ISource
+        key_types.push_back(type);
+    }
+
+    /// `joinGet` returns a single column, so request the columns one by one.
```
🕵 Fixed in `7a56eace474`. `getBlockByKeys` now takes a list of keys instead of one key, builds a multi-row key block, and looks up every key and every column under a single `StorageJoin` read lock taken in `getBlockByKeys` itself (calling `join->joinGet` directly instead of the public `StorageJoin::joinGet`, which would re-take the lock per call). The keys are processed in batches of `DEFAULT_BLOCK_SIZE`, because `joinGet` returns at most `max_joined_block_rows` rows at a time, and the result size is verified against the number of keys. `MGET` issues one request for all of its keys, so a single `MGET`/`HMGET` can no longer mix values from different states of the table.

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,405 @@
+{
+    /// Resolve the table again for every request: a session must not keep serving data from a table
+    /// that has been dropped, renamed, or recreated in the meantime.
+    auto table = DatabaseCatalog::instance().getTable(StorageID{mapping.clickhouse_db, mapping.clickhouse_table}, query_context);
```
🕵 Fixed in ac5ba903589b: `resolveTable` now takes `IStorage::lockForShare` (with `lock_acquire_timeout`, like a regular read) right after resolving the table, and the handler holds the returned `TableLockHolder` for the whole command, so a concurrent `DROP TABLE` / `DETACH TABLE` cannot commit while a `GET`/`MGET`/`HGET`/`HMGET` is still reading the table.

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,405 @@
+            "Key column {} of table {} configured for Redis database {} must be of type String or FixedString, but it is {}",
+            mapping.key_column, mapping.clickhouse_table, db_, sample_block.getByName(mapping.key_column).type->getName());
+
+    if (mapping.db_type == RedisProtocol::DBType::STRING && !sample_block.has(mapping.value_column))
```
🕵 Fixed in ac5ba903589b: `SELECT` on a `string` database now runs the same `joinGetOrNull` return-type check as the lookups, by calling `getBlockByKeys` with no keys (it validates the return types before reading anything). A mapping whose `value_column` cannot be wrapped in `Nullable` (e.g. `Array`) is now rejected by `SELECT` with the "cannot be used for get requests" exception instead of failing on the first `GET`. Covered by the new `test_unrepresentable_value_column` integration test.

alexey-milovidov commented on src/Server/RedisHandler.cpp:
```diff
@@ -0,0 +1,422 @@
+                throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "GET command can only be applied to a database of type string");
+
+            auto [table, table_lock] = resolveTable(db, mapping);
+            query_context->checkAccess(AccessType::SELECT, table->getStorageID(), Strings{mapping.key_column, mapping.value_column});
```
🕵 Fixed in 7c6123a062a3. `RedisHandler::validateTable` now refuses to serve a table the current user has a row policy on (`ACCESS_DENIED`, checked on `SELECT` and on every command, since the table and the mapping are resolved per command), instead of returning rows the policy hides. Added `test_row_policy_is_rejected`.

alexey-milovidov commented on src/Server/RedisProtocolRequest.h:
```diff
@@ -0,0 +1,359 @@
+        else if (cmd == Command::CLIENT)
+            command = CommandType::CLIENT;
+        else
+            throw Exception(ErrorCodes::UNKNOWN_PACKET_FROM_CLIENT, "Unknown command: {}", cmd);
```
🕵 Fixed in 7c6123a062a3. `RedisRequest::deserialize` now reads the whole RESP array (header, command name and all arguments) before anything is interpreted, and `parse` classifies the command afterwards, so an unknown command or a wrong-arity command consumes its arguments and the connection stays in sync. A framing error — after which the position in the stream is unknown — now closes the connection instead of replying and continuing. The total size of the arguments of one command is capped by `MAX_COMMAND_SIZE`. Added `test_error_does_not_desynchronize_connection`.

alexey-milovidov commented on src/Storages/StorageJoin.cpp:
```diff
@@ -1106,6 +1109,95 @@ class JoinSource final : public ISource
+
+        /// Request a Nullable result (as `joinGetOrNull` does), so that a missing key is
+        /// distinguishable from a key present with a default value.
+        auto return_type = joinGetCheckAndGetReturnType(key_types, name, /*or_null=*/ true);
```
🕵 Fixed in 7c6123a062a3. `StorageJoin::getBlockByKeys` now asks `joinGetCheckAndGetReturnType` for the plain type when the column is `LowCardinality` and wraps it with `makeNullableOrLowCardinalityNullableSafe`, so the lookup returns `LowCardinality(Nullable(T))`. The change is kept in the new lookup path rather than in `joinGetCheckAndGetReturnType` itself, to avoid changing the behaviour of the `joinGetOrNull` function in the same commit. Added `test_low_cardinality_value_column`.

alexey-milovidov commented on src/Server/RedisProtocolRequest.h:
```diff
@@ -0,0 +1,361 @@
+        Int64 command_len = reader.readInteger();
+        if (command_len < 1)
+            throw Exception(ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT, "Wrong command length: {}", command_len);
+        if (command_len > MAX_ARRAY_SIZE)
```
🕵 Fixed. The number of elements of a single command is now limited on its own (`MAX_COMMAND_ELEMENTS` = 64 Ki), and every element is additionally charged a fixed `COMMAND_ELEMENT_MEMORY_OVERHEAD` (128 bytes) against `MAX_COMMAND_SIZE`, so a short argument is accounted for what it really costs rather than for its one payload byte. The example request `*1048576\r\n$4\r\nMGET\r\n …` is now rejected from the array header alone, before a single key is read. `MGET` also no longer copies the keys out of the request — `MGetRequest` is a view over `RedisRequest::arguments` — so one materialization of the keys is gone.

Worst case for an accepted command is now ~64 Ki elements: the payload it holds is bounded by `MAX_COMMAND_SIZE` and is proportional to what was actually sent on the wire, and the per-element containers add ~10 MiB rather than hundreds of MiB.

Regression test: `test_many_tiny_keys_are_rejected` sends the raw RESP command with a million tiny keys and asserts it is refused, and then asserts that a 1024-key `MGET` is still answered normally. All 15 tests of `test_redis_protocol` pass locally.

alexey-milovidov commented:
🕵 Merged `master` and fixed the remaining AI-review blocker.

**Bounded the memory of a single command.** `MAX_COMMAND_SIZE` only counted the payload bytes of the arguments, so a command with a million one-byte keys was a few `MiB` on the wire but hundreds of `MiB` of server memory (held as `RedisRequest::arguments`, as `MGetRequest::keys` and as the `std::vector<std::vector<Field>>` of the lookup). Now:

- a command is limited to `MAX_COMMAND_ELEMENTS` (64 Ki) elements — the example `*1048576\r\n$4\r\nMGET\r\n …` is rejected from the array header, before any key is read;
- every element is additionally charged `COMMAND_ELEMENT_MEMORY_OVERHEAD` (128 bytes) against `MAX_COMMAND_SIZE`, so short arguments are accounted for what they really cost;
- `MGET` no longer copies the keys out of the request — `MGetRequest` is a view over `RedisRequest::arguments`.

New regression test `test_many_tiny_keys_are_rejected`: the raw RESP command with a million tiny keys is refused, and a 1024-key `MGET` is still answered normally. `ninja clickhouse` builds cleanly and all **15** tests of `test_redis_protocol` pass locally.

The only red on the previous head was `Stress test (amd_debug)` with `Logical error: 'Equal values are not contiguous within the range assumed to be sorted'`. It is unrelated to this pull request: the same failure appears on `master` and on ~30 unrelated pull requests over the last month, and it is tracked by https://github.com/ClickHouse/ClickHouse/issues/115455.
